pkg/controller/worker/reconciler.go

```diff
@@ -0,0 +1,174 @@
+	}
+
+	// Deletion flow
+	if worker.DeletionTimestamp != nil {
+		hasFinalizer, err := extensionscontroller.HasFinalizer(worker, FinalizerName)
+		if err != nil {
+			r.logger.Error(err, "Could not instantiate finalizer deletion")
+			return reconcile.Result{}, err
+		}
+		if !hasFinalizer {
+			r.logger.Info("Deleting worker causes a no-op as there is no finalizer.", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+			return reconcile.Result{}, nil
+		}
+
+		operationType := gardencorev1alpha1helper.ComputeOperationType(worker.ObjectMeta, worker.Status.LastOperation)
+		if err := r.updateStatusProcessing(r.ctx, worker, operationType, "Deleting the worker"); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		r.logger.Info("Starting the deletion of worker", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+		if err := r.actuator.Delete(r.ctx, worker, cluster); err != nil {
+			msg := "Error deleting worker"
+			utilruntime.HandleError(r.updateStatusError(r.ctx, extensionscontroller.ReconcileErrCauseOrErr(err), worker, operationType, msg))
+			r.logger.Error(err, msg, "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+			return extensionscontroller.ReconcileErr(err)
+		}
+
+		msg := "Successfully deleted worker"
+		r.logger.Info(msg, "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+		if err := r.updateStatusSuccess(r.ctx, worker, operationType, msg); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		r.logger.Info("Removing finalizer.", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+		if err := extensionscontroller.DeleteFinalizer(r.ctx, r.client, FinalizerName, worker); err != nil {
+			r.logger.Error(err, "Error removing finalizer from Worker", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// Reconcile flow
+	if err := controller.EnsureFinalizer(r.ctx, r.client, FinalizerName, worker); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	operationType := gardencorev1alpha1helper.ComputeOperationType(worker.ObjectMeta, worker.Status.LastOperation)
+	if err := r.updateStatusProcessing(r.ctx, worker, operationType, "Reconciling the worker"); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.actuator.Reconcile(r.ctx, worker, cluster); err != nil {
+		msg := "Error reconciling worker"
+		utilruntime.HandleError(r.updateStatusError(r.ctx, extensionscontroller.ReconcileErrCauseOrErr(err), worker, operationType, msg))
+		r.logger.Error(err, msg, "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+		return extensionscontroller.ReconcileErr(err)
+	}
+
+	msg := "Successfully reconciled worker"
+	r.logger.Info(msg, "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+	if err := r.updateStatusSuccess(r.ctx, worker, operationType, msg); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
```

Comment:
Since this controller deploys RBAC manifests to the Shoot cluster, don't you think it's necessary to requeue reconciliation every x minutes like we do here https://github.com/gardener/gardener-extensions/blob/8aa82e30ce72025cc664a96338f7adf3b67a8cac/pkg/controller/extension/reconciler.go#L167?

Reply:
I would first try to go ahead without this as the resources in the shoot shall be managed by the gardener-resource-manager in the future anyways. And, as the worker controller does quite a lot of things I would start to not do the reconciliation too often. Of course, there is the risk that end-users delete these resources and that it might take some time until they get re-created (in the next `Worker` reconciliation), however, then they a) have harmed themselves, and b) I consider the risk quite low. Let's get some experience first - if we see that too many end-users tamper with these RBAC roles then let's do it the way you suggested. Otherwise, we hopefully have the gardener-resource-manager in a few weeks and don't need to think about these this anymore.